ports/unix/mpconfigport.h

```diff
@@ -362,7 +362,12 @@ struct _mp_bluetooth_nimble_malloc_t;
 #define MICROPY_END_ATOMIC_SECTION(x) (void)x; mp_thread_unix_end_atomic_section()
 #endif
 
-#define MICROPY_EVENT_POLL_HOOK mp_hal_delay_us(500);
+#define MICROPY_EVENT_POLL_HOOK \
+    do { \
+        extern void mp_handle_pending(bool); \
+        mp_handle_pending(true); \
+        mp_hal_delay_us(500); \
```

Comment:
The GIL should be released during the delay to allow other threads to run when GIL and threads are enabled.

Reply:
It might be already? If not, that's probably a separate issue to this one - the `mp_hal_delay_us(500)` is what's already there as `MICROPY_EVENT_POLL_HOOK` before my addition of the `mp_handle_pending`

Reply:
Ah nope, it looks like unit `mp_hal_delay_us` is just running posix `usleep`, so there's no GIL wrapper in there. I'm not sure what the implications would be of releasing the GIL in MICROPY_EVENT_POLL_HOOK though, I'm not sure that's actually safe?

Reply:
The Unix port enables threading but, by default, without the GIL enabled (see unix/Makefile:133). Unlike CPython, concurrent threads are therefore running the VM in parallel and are not able to safely mutate global state (i.e. modify the same lists), but with the benefit (of course) that two threads can actually be executing code simultaneously.

When code is run by the scheduler, it's either inline in the VM (as if the call to the scheduled function was invokved directly between two instructions in the user's code) (so the GIL will already be held), or due to the EVENT_HOOK, in which case (if the GIL is enabled) we must already be holding it (i.e. you must not call MICROPY_EVENT_POLL_HOOK if you have explicitly released the GIL).

So the assumption is that the GIL protects modification of Python state, and so as long as the thing calling MICROPY_EVENT_POLL_HOOK isn't mid-modification of some structure, then it should be safe to unlock the GIL, so I agree with @dlech here. (Although see below, I'm not sure this is necessarily the right place to do it) And the STM32 port does exactly that in it's implementation of MICROPY_EVENT_POLL_HOOK: releasing the GIL and explicitly yielding to another thread. Maybe this could be fixed in a different PR?

However, back to mp_handle_pending -- almost nothing in the unix port actually calls MICROPY_EVENT_POLL_HOOK, this is really just a placeholder implementation. One that always gets me is that while sitting at the REPL the scheduler will not run any tasks. (On STM32 the REPL polls the UART and in between polling it calls MICROPY_EVENT_POLL_HOOK, whereas on Unix, it's just doing a blocking `read()` on stdin).

The only reason that other threads continue to run in the background while at the REPL is that the call do read() is done via `MP_HAL_RETRY_SYSCALL()` which releases the GIL around the `read()`. So that makes me think that the call to `usleep()` (inside `mp_hal_delay_{ms,us}`) should release/re-acquire the GIL. This would also fix another issue on the unix port, which is that when the GIL is enabled, a `time.sleep()`'ing thread blocks all other execution.

I was wondering how this issue with the schduler on Unix came up, given that nothing on Unix uses it (and based on the above isn't because it's at the REPL), but now I see in #6438 that it now requires that the scheduler runs while it polls for the NimBLE semaphore to become ready, which is why this fix is necessary to make that work.

So..... I think that was a lot of words to say:
 - @dlech is correct that this should release the GIL, but unrelated to the PR (and only makes a difference in a non-default configuration), and fixing it should probably be done in `mp_hal_delay_{ms,us}` instead.
 - To actually notice this PR doing anything today (i.e. seeing scheduled tasks running at the REPL) requires other changes to the Unix port (e.g. making the stdin read() also call MICROPY_EVENT_POLL_HOOK).
 - This PR as it is is probably fine to merge as-is.

Reply:
I think we first need to fix other issues with `micropython.schedule()` on unix, then this would eventually fall out of it.  Eg get a reliable `machine.Timer()` working.